api/envoy/config/listener/v3/listener_components.proto

```diff
@@ -357,3 +359,164 @@ message ListenerFilter {
+message DirectSourceIPInput {
+}
+
+// Matcher input that allows matching connections by by the source port.
+message SourcePortInput {
+}
+
+// Matcher input that allows matching connections by by the connection source IP type.
+// Specifies the connection source IP match type. The values include:
+//
+// * ``any`` - any connection source,
+// * ``local`` - matches a connection originating from the same host,
+// * ``external`` - matches a connection originating from a different host
+message SourceTypeInput {
+}
+
+// Matcher input that allows matching connections by by the server names of a new connection, when detected by one of the listener filters.
+message SNIInput {
+}
+
+// Matcher input that allows matching connections by by the transport protocol.
+// This value will be compared against the transport protocol of a new connection, when
+// it's detected by one of the listener filters.
+//
+// Suggested values include:
+//
+// * ``raw_buffer`` - default, used when no transport protocol is detected,
+// * ``tls`` - set by :ref:`envoy.filters.listener.tls_inspector <config_listener_filters_tls_inspector>`
+//   when TLS protocol is detected.
+message TransportProtocolInput {
+}
+
+// Matcher input that allows matching connections by by the application protocol (e.g. ALPN for TLS protocol).
+// This values will be compared against the application
+// protocols of a new connection, when detected by one of the listener filters.
+//
+// Suggested values include:
+//
+// * ``http/1.1`` - set by :ref:`envoy.filters.listener.tls_inspector
+//   <config_listener_filters_tls_inspector>` and :ref:`envoy.filters.listener.http_inspector
+//   <config_listener_filters_http_inspector>`,
+// * ``h2`` - set by :ref:`envoy.filters.listener.tls_inspector <config_listener_filters_tls_inspector>` and :ref:`envoy.filters.listener.http_inspector <config_listener_filters_http_inspector>`
+//
+// .. attention::
+//
+//   Currently, :ref:`TLS Inspector <config_listener_filters_tls_inspector>` provides
+//   application protocol detection based on the requested
+//   `ALPN <https://en.wikipedia.org/wiki/Application-Layer_Protocol_Negotiation>`_ values.
+//
+//   However, the use of ALPN is pretty much limited to the HTTP/2 traffic on the Internet,
+//   and matching on values other than ``h2`` is going to lead to a lot of false negatives,
+//   unless all connecting clients are known to use ALPN.
+message ApplicationProtocolInput {
+}
+
+// Matches a specific port against a set of possibly overlapping port ranges.
+// TODO(kuat) Move to cncf/xds.
+message PortMatcher {
+  // Specifies an optional list of port ranges and a match action.
+  message PortRangeMatcher {
+    // Optional set of port ranges. An empty set matches any port.
+    repeated type.v3.Int32Range ranges = 1;
+
+    // Match action to apply when the port is within one of the port ranges.
+    xds.type.matcher.v3.Matcher.OnMatch on_match = 2;
```

Comment:
I'm not very familiar with the new matching API, but it seems odd to me that every matcher has to supply its own on_match field? Shouldn't that be handled outside the specific matcher somehow? cc @snowp

Reply:
If this is implementing a match tree then you would specify the OnMatch within this proto, similar to how ExactMatchMap works: https://github.com/cncf/xds/blob/main/xds/type/matcher/v3/matcher.proto#L103-L106

If we want this to be able to say "if the port falls within range A resolve match action X, if port falls within range B resolve match action Y" we would need this kind of structure

Looking at this structure in more detail I would think that we could perhaps simplify it by having a PortRangeMatcher (which maps a list of ranges to an OnMatch) and use the existing ExactMatch mechanism to implement the port_map. You'd then use a `on_no_match` on the exact matcher to describe that if it's not found in that map it would attempt the next matcher, which would be the ip ranges one

Reply:
@snowp Good point about using on_no_match on exact port matcher. It's not as expressive with respect to exclusivity but that is also fairly contrived use case, so it's not worth tackling that first.

Reply:
Updated to simplify.